Refactor methods for rendering documents, add excerpts and fix content attributes

nott opened the pull request:
Note: the scope of this PR has changed.
- This PR adds Jekyll-like excerpts. Please see [Jekyll docs](https://jekyllrb.com/docs/posts/#post-excerpts) for details. The previous attempt to do it was #145 .
- Fixes content attribute added in #146 - this attribute shouldn't include extended layout, just the document.
- Changes the way we split attributes and content.
- Fixes bug with two sequential rendering when `extends` is not defined.
- Adds a layout cache (#120).

Method `as_html` was replaces with `render` which now has side-effects.

nott commented:
General notice:

* on the one hand writing tests with fxitures/target
* however it seems that they it will be hard to support these tests in future
* it's not very convenient to cover `Document::as_html` with unit tests when this function accepts paths (for `source` and `layouts_path`) instead of file contents

We should probably move reading `source` and `layout` into separate functions and cover all the functions with unit tests.

nott commented:
I'll probably extend the scope of this PR in a day or two.

nott commented:
@johannhof could you please review this?

johannhof commented on src/document.rs:
```diff
@@ -264,17 +258,79 @@ impl Document {
+                  -> Result<String> {
+        let content_html = try!(self.render_html(&self.content, context, source));
+        self.attributes.insert("content".to_owned(), Value::Str(content_html.clone()));
+        context.set_val("content", Value::Str(content_html.to_owned()));
```
please use `clone` instead of `to_owned` here, too

johannhof commented on src/document.rs:
```diff
@@ -264,17 +258,79 @@ impl Document {
+                          context: &mut Context,
+                          source: &Path,
+                          default_excerpt_separator: &str)
+                          -> Result<String> {
```
This doesn't have to return a `String`

johannhof commented on src/document.rs:
```diff
@@ -264,17 +258,79 @@ impl Document {
+                    .filter(|chunk| !chunk.is_empty())
+                    .next() // first non-empty chunk
+                    .map(|s| s.to_string())
+                    .unwrap_or(self.content.to_string())
```
It's not a huge thing but you can save on allocation and code if you make `excerpt` a `&str` instead of an owned String. You'd have to move it into a block for creating `excerpt_html` to avoid the dangling reference to `self.attributes`, though.

johannhof replied:
Hm maybe that comment wasn't really clear, here's how I'd implement it:
```
    /// Renders excerpt and adds it to attributes of the document.
    pub fn render_excerpt(&mut self,
                          context: &mut Context,
                          source: &Path,
                          default_excerpt_separator: &str)
                          -> Result<()> {
        let excerpt_html = {
            let excerpt_attr = self.attributes
                .get("excerpt")
                .and_then(|attr| attr.as_str());

            let excerpt_separator = self.attributes
                .get("excerpt_separator")
                .and_then(|attr| attr.as_str())
                .unwrap_or(default_excerpt_separator);

            let excerpt = if let Some(excerpt_str) = excerpt_attr {
                excerpt_str
            } else if excerpt_separator.is_empty() {
                ""
            } else {
                self.content.split(excerpt_separator)
                        .filter(|chunk| !chunk.is_empty())
                        .next() // first non-empty chunk
                        .unwrap_or(&self.content)
            };

            try!(self.render_html(excerpt, context, source))
        };

        self.attributes.insert("excerpt".to_owned(), Value::Str(excerpt_html));
        Ok(())
    }
```

johannhof commented on src/document.rs:
```diff
@@ -264,17 +258,79 @@ impl Document {
+    /// Side effects:
+    ///
+    /// * content is inserted to the attributes of the document
+    /// * content is inserted to context
```
another side effect is that the layout is inserted into the `layouts_cache`

nott replied:
nice catch

johannhof commented on src/cobalt.rs:
```diff
@@ -48,6 +49,7 @@ pub fn build(config: &Config) -> Result<()> {
 
     let layouts = source.join(&config.layouts);
     let layouts = layouts.as_path();
+    let mut layouts_cache = HashMap::new();
```
thanks for adding this

nott commented:
@johannhof thank you for the feedback!
i'll address all your comments later today or tomorrow

johannhof commented:
Sounds great, thanks!

nott commented:
> Sorry for the delay, I had to think about this. As you know I'm not a big fan of rendering twice, but I think you're right, there's no other way to do this if we want excerpts.

We can't avoid rendering twice indeed. I was thinking about lazy rendering on first usage, but it complicates the code. I don't think it makes sense to add this feature having the existing infrastructure... But it's a good project for future that can improve both liquid-rust and cobalt.

nott commented:
@johannhof all your comments have been addressed.

There was one more additional change in `render_excerpt`:

Instead of
```
                self.content.split(excerpt_separator)
                        .filter(|chunk| !chunk.is_empty())
                        .next() // first non-empty chunk
                        .unwrap_or(&self.content)
```
I used
```
self.content.split(excerpt_separator).next().unwrap_or(&self.content)
```

Skipping leading empty chunk was imported before 8af758b was merged. But as soon as 8af758b is merged, skipping empty chunks doesn't make much sense (jekyll also doesn't do it and it's contr-intuitive to skip them when using custom separators like `<!-- more -->`)

johannhof commented:
Great work, thank you!

nott commented:
i've just noticed one `clone` that should have been removed... fixed
